### DISCUSSION TITLE
Stop Analysis function in core API

### DISCUSSION THREAD

**Discussion by sashokbg** [pmd/pmd#4743] [Q&A] [answered]
Hello, within the PmdAnalysis class in the core library there is a "performAnalysis" method, but I do not see any API methods that can cancel / stop the current analysis. 

Did I look at the wrong place or it is not yer developed ? 

This feature is needed when we use PMD within a plugin such as PMD-Intellij. The IDE has hooks for "progress start" and "progress end" and "progress cancel" events that should be respected if possible.

Thank you for your help

References:

https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/PmdAnalysis.java#L360
https://github.com/amitdev/PMD-Intellij

**Comment by jsotuyod** [comment, marked as answer]
@sashokbg there is currently no public API to do this. What you can do at the moment is simply interrupting the thread used to launch the PMD analysis. Even if the analysis is multi-threaded it will completely shutdown the analysis.

We can't expose this as an API, as it assumes the current thread can be discarded, or requires to run everything in a worker thread in the off-scenario where this is actually needed… so it seems better to leave the consumer of such scenario to do the wrapping.

Alternatively we *could* drop the mono thread processor, have everything run on an executor service, even if it's only 1 thread, and expose an API to shutdown the executor, but that would fail to stop the analysis unless it has completely finalized the setup step, including figuring out the exhaustive list of files to be analyzed by each subset of rules… but unlikely to happen in the short term as we focus on PMD 7.0.0 final release.

**Reply by adangel** [reply]
Yes, we need a API to support cancellation - which we don't have yet. We only have progress reporting (see #3828), but we didn't implement something like a `ProgressIndicator` which has a `isCancelled` method.

**Reply by sashokbg** [reply]
Should I open an issue so that it gets tracked there and eventually developed by some motivated folk :) ?

**Reply by oowekyala** [reply]
The original comment of #3828 describes a way to implement cancelling by throwing an exception in a FileAnalysisListener. Can that be used as is by the intellij plugin @sashokbg?
